shared/lib/multichain/chain-agnostic-permission.ts

```diff
@@ -0,0 +1,286 @@
+import {
+  Caip25CaveatValue,
+  InternalScopesObject,
+  InternalScopeObject,
+} from '@metamask/chain-agnostic-permission';
+import {
+  CaipAccountId,
+  CaipChainId,
+  parseCaipAccountId,
+  parseCaipChainId,
+} from '@metamask/utils';
+import { isEqualCaseInsensitive } from '../../modules/string-utils';
+
+// Type for an internal account with minimum required properties
+export interface InternalAccount {
+  address: string;
+  scopes: string[];
+}
+
+/**
+ * Gets all permitted accounts from a CAIP-25 caveat
+ * This extracts all account IDs from both required and optional scopes
+ * and returns a unique set.
+ *
+ * @param caip25CaveatValue - The CAIP-25 caveat value to extract accounts from
+ * @returns Array of unique account IDs
+ */
+export function getAllPermittedAccounts(
+  caip25CaveatValue: Caip25CaveatValue,
+): CaipAccountId[] {
+  const permittedAccounts = new Set<CaipAccountId>();
+
+  // Extract accounts from required scopes
+  Object.values(caip25CaveatValue.requiredScopes).forEach(({ accounts }) => {
+    accounts.forEach((account) => {
+      permittedAccounts.add(account as CaipAccountId);
+    });
+  });
+
+  // Extract accounts from optional scopes
+  Object.values(caip25CaveatValue.optionalScopes).forEach(({ accounts }) => {
+    accounts.forEach((account) => {
+      permittedAccounts.add(account as CaipAccountId);
+    });
+  });
+
+  return Array.from(permittedAccounts);
+}
+
+/**
+ * Gets all scopes (chain IDs) from a CAIP-25 caveat
+ * This extracts all scopes from both required and optional scopes
+ * and returns a unique set.
+ *
+ * @param caip25CaveatValue - The CAIP-25 caveat value to extract scopes from
+ * @returns Array of unique scope strings (chain IDs)
+ */
+export function getAllScopes(
+  caip25CaveatValue: Caip25CaveatValue,
+): CaipChainId[] {
+  // Extract scopes from required and optional
+  const requiredScopes = Object.keys(caip25CaveatValue.requiredScopes);
+  const optionalScopes = Object.keys(caip25CaveatValue.optionalScopes);
+
+  // Combine and return unique scopes
+  return Array.from(
+    new Set([...requiredScopes, ...optionalScopes]),
+  ) as CaipChainId[];
+}
+
+/**
+ * Updates the list of permitted accounts in a CAIP-25 caveat with the given accounts
+ *
+ * @param caip25CaveatValue - The original CAIP-25 caveat value
+ * @param updatedAccounts - The array of accounts to set
+ * @returns The updated CAIP-25 caveat value with new accounts
+ */
+export function updatePermittedAccounts(
```

Comment:
isn't this effectively the same as the setPermittedAccounts helper we already have?

Reply:
yes lol and it doesn't appear to be used...